### PR TITLE
feat: rewrite ratelimit to be more extensible

### PR DESCRIPTION
**Pull request by BlankParticle**
## What does this PR do?

Rate limit system has been re written to be more extensible and usable with all trpc routes. 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Required

- [ ] Read [Contributing Guide](https://github.com/un/inbox/blob/main/CONTRIBUTING.md)
- [ ] Self-reviewed my own code
- [ ] Tested my code in a local environment
- [ ] Commented on my code in hard-to-understand areas
- [ ] Checked for warnings, there are none
- [ ] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the UnInbox Docs if changes were necessary

### CODE REVIEW

**Comment by BlankParticle** [comment]
* **#541** <a href="https://app.graphite.dev/github/pr/un/inbox/541?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @BlankParticle and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: apps/platform/trpc/ratelimit.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   */
+  namespace?: RatelimitConfig['namespace'];
+  /**
+   * Function to create the identifier for the rate limit
+   * @default ipIdentifier
+   */
+  createIdentifier?: (ctx: TrpcContext) => string;
+};
+
+export const createRatelimit = ({
```
Comment (McPizza0) [inline apps/platform/trpc/ratelimit.ts:41]: rename to ratelimiter to make it clearer/cleaner

so in usage its `.use(ratelimiter({ limit: 30 })

**Review by McPizza0 — COMMENTED** [review]
should use namespaces so we know which type of limits they are
`signup, signin, convos etc.`

File: apps/platform/trpc/ratelimit.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  limit,
+  duration = '1h',
+  namespace = 'public-functions',
+  createIdentifier = ipIdentifier
+}: CreateRateLimiterConfig) =>
+  trpcContext.middleware(async ({ ctx, path, next }) => {
+    let limiter = CACHE_MAP.get(path);
```
[the remark this answers is not in the document]
Comment (BlankParticle) [reply apps/platform/trpc/ratelimit.ts:48]: renaming to `cachedLimiters`

File: apps/platform/trpc/routers/authRouter/passwordRouter.ts [hunk 3]
```diff
@@ -74,8 +76,9 @@ export const passwordRouter = router({
       return { success: true };
     }),
 
-  signUpWithPassword2FA: publicRateLimitedProcedure.signUpWithPassword
+  signUpWithPassword2FA: publicProcedure
     .unstable_concat(turnstileProcedure)
+    .use(ratelimiter({ limit: 10, namespace: 'signUp.password' }))
```
Comment (McPizza0) [inline apps/platform/trpc/routers/authRouter/passwordRouter.ts:81]: duplicate name @BlankParticle

Comment (BlankParticle) [reply apps/platform/trpc/routers/authRouter/passwordRouter.ts:81]: one of the route is not used and is kept for not having issues in the old Nuxt app

File: apps/platform/trpc/routers/authRouter/passwordRouter.ts [hunk 4]
```diff
@@ -307,8 +311,9 @@ export const passwordRouter = router({
       });
     }),
 
-  signIn: publicRateLimitedProcedure.signInWithPassword
+  signIn: publicProcedure
     .unstable_concat(turnstileProcedure)
+    .use(ratelimiter({ limit: 20, namespace: 'signIn.password' }))
```
Comment (McPizza0) [inline apps/platform/trpc/routers/authRouter/passwordRouter.ts:316]: duplicate name @BlankParticle

File: apps/platform/trpc/routers/authRouter/passwordRouter.ts (hunk 3 above)
Comment (McPizza0) [reply apps/platform/trpc/routers/authRouter/passwordRouter.ts:81]: ah ok

**Review by McPizza0 — APPROVED** [review]

**Comment by BlankParticle** [comment]
### Merge activity

* **Jul 4, 2:42 PM EDT**: @BlankParticle merged this pull request with [Graphite](https://app.graphite.dev/github/pr/un/inbox/541).

**Pull request merged**